### PR TITLE
Feat(Genotype) Added compatibility for balsamic cases

### PR DESCRIPTION
**Pull request by Vince-janv**
## Description
Currently the genotype api code is only compatible with mip cases. This needd to be changed for [this issue](https://github.com/Clinical-Genomics/BALSAMIC/issues/882).
### Added

- Possibility to use the genotype apis on balsamic cases.

### How to prepare for test
- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh feat-balsamic-genotype`
- [x] Make sure that the case `gladyeti` exists in housekeeper stage
- [x] Make sure that the file `/home/proj/stage/housekeeper-bundles/justfrog/2022-06-01/justfrog_gatkcomb.bcf` exists.
- [x] Make sure that the case  `profowl` exists in housekeeper stage
- [x] Make sure that the file `/home/proj/stage/housekeeper-bundles/profowl/2022-06-03/SNV.germline.tumor.dnascope.vcf.gz` exists in housekeeper stage and has the tag `genotype`
- [x] reheader `/home/proj/stage/housekeeper-bundles/profowl/2022-06-03/SNV.germline.tumor.dnascope.vcf.gz` to the correct header with `bcftools reheader -s change_sample_names.txt SNV.germline.normal.dnascope.vcf.gz -o SNV.germline.normal.dnascope.changednamed.vcf.gz`

### How to test
- [x] run `cg upload genotype gladyeti`
- [x] run `cg upload genotype profowl`

### Expected test outcome
- [x] `gladyeti` is uploaded without error
- [x] `profowl` is uploaded without error
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [x] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by Vince-janv** [comment]
### Loading balsamic case:
Screenshot 2022-06-21 at 12 55 21

### Loading Mip case
Screenshot 2022-06-21 at 13 08 31

File: cg/meta/upload/gt.py [hunk 1]
```diff
@@ -45,16 +47,44 @@ def data(self, analysis_obj: models.Analysis) -> dict:
...
+        data = {"bcf": hk_bcf.full_path}
+        if analysis_obj.pipeline in [Pipeline.BALSAMIC, Pipeline.BALSAMIC_UMI]:
+            data["samples_sex"] = self._get_sample_sex_balsamic(
+                case_obj=analysis_obj.family, hk_version=hk_version
+            )
+        elif analysis_obj.pipeline == Pipeline.MIP_DNA:
+            data["samples_sex"] = self._get_sample_sex_mip(
+                case_obj=analysis_obj.family, hk_version=hk_version
+            )
+        return data
```
Comment (patrikgrenfeldt) [inline cg/meta/upload/gt.py:59]:
```suggestion
        else:
          # raise error
        return data
```

File: cg/meta/upload/gt.py [hunk 2]
```diff
@@ -45,16 +47,44 @@ def data(self, analysis_obj: models.Analysis) -> dict:
...
+        self, case_obj: models.Family, hk_version: housekeeper_models.Version
+    ) -> dict:
+        samples_sex = {}
+        for link_obj in case_obj.links:
+            if link_obj.sample.is_tumour:
+                continue
+            sample_id = link_obj.sample.internal_id
+            samples_sex[sample_id] = {
+                "pedigree": link_obj.sample.sex,
+                "analysis": "unknown",
```
Comment (patrikgrenfeldt) [inline cg/meta/upload/gt.py:85]:
```suggestion
                "analysis": Gender.unknown,
```

File: cg/meta/upload/gt.py [hunk 3]
```diff
@@ -45,16 +47,44 @@ def data(self, analysis_obj: models.Analysis) -> dict:
...
+            data["samples_sex"] = self._get_sample_sex_balsamic(
+                case_obj=analysis_obj.family, hk_version=hk_version
+            )
+        elif analysis_obj.pipeline == Pipeline.MIP_DNA:
+            data["samples_sex"] = self._get_sample_sex_mip(
+                case_obj=analysis_obj.family, hk_version=hk_version
+            )
+        return data
+
+    def _get_sample_sex_mip(
```
Comment (patrikgrenfeldt) [inline cg/meta/upload/gt.py:61]:
```suggestion
    def _get_samples_sex_mip(
```
also for balsamic

File: cg/meta/upload/gt.py [hunk 4]
```diff
@@ -67,7 +97,7 @@ def analysis_sex(self, qc_metrics_file: Path) -> dict:
     def get_bcf_file(self, hk_version_obj: housekeeper_models.Version) -> housekeeper_models.File:
         """Fetch a bcf file and return the file object"""
 
-        bcf_file = self.hk.files(version=hk_version_obj.id, tags=["snv-gbcf"]).first()
+        bcf_file = self.hk.files(version=hk_version_obj.id, tags=["genotype"]).first()
```
Comment (patrikgrenfeldt) [inline cg/meta/upload/gt.py:100]: get the file not being the index

**Review by patrikgrenfeldt — APPROVED** [review]
I approve 👍

File: cg/meta/upload/gt.py (hunk 4 above)
Comment (Vince-janv) [reply cg/meta/upload/gt.py]: Can you have a look at the new solution for this comment

**Review by patrikgrenfeldt — APPROVED** [review]
Great

File: cg/meta/upload/gt.py [hunk 5]
```diff
@@ -96,8 +92,19 @@ def analysis_sex(self, qc_metrics_file: Path) -> dict:
 
     def get_bcf_file(self, hk_version_obj: housekeeper_models.Version) -> housekeeper_models.File:
         """Fetch a bcf file and return the file object"""
-
-        bcf_file = self.hk.files(version=hk_version_obj.id, tags=["genotype"]).first()
+        bcf_file = None
+        genotype_files = self.hk.files(version=hk_version_obj.id, tags=["genotype"]).all()
```
Comment (Mropat) [inline cg/meta/upload/gt.py]:
```suggestion
        genotype_files: list = self.hk.files(version=hk_version_obj.id, tags=["genotype"]).all()
```

File: cg/meta/upload/gt.py [hunk 6]
```diff
@@ -96,8 +92,19 @@ def analysis_sex(self, qc_metrics_file: Path) -> dict:
...
+        while not bcf_file or not genotype_files:
```
Comment (Mropat) [inline cg/meta/upload/gt.py]: Think this is more readable as a for loop over genotype files!

**Comment by Vince-janv** [comment]
### Deployed on stage:

> Logging deploy ...
> Getting deployer... done.
> Getting last commit message and SHA... done.
> Getting version of deploy scripts... /home/proj/production/servers/resources/hasta.scilifelab.se
> done.
> Log deploy... done.
> Running --version on each Clinical-Genomics developed tool to verify that all requirements are met:
> vogue 1.1.1 requires multiqc, which is not installed.
> cg-fluffy 3.0.0 requires slurmpy, which is not installed.
> typer 0.3.2 has requirement click<7.2.0,>=7.1.1, but you have click 6.7.
> demux 5.4.5 has requirement importlib-metadata<4.0.0,>=3.1.1, but you have importlib-metadata 2.1.2.
> cg-hermes 1.6.14 has requirement cgmodels<0.5.0,>=0.4.0, but you have cgmodels 0.10.1.
> /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
>   "viewonly=True." % (", ".join(sorted(non_viewonly)))
> cg, version 22.68.0

### Deployed in prod:

> Logging deploy ...
> Getting deployer... done.
> Getting last commit message and SHA... done.
> Getting version of deploy scripts... /home/proj/production/servers/resources/hasta.scilifelab.se
> done.
> Log deploy... done.
> Running --version on each Clinical-Genomics developed tool to verify that all requirements are met:
> balsamic 6.0.0 requires colorclass, which is not installed.
> argcomplete 1.12.1 has requirement importlib-metadata<3,>=0.23, but you have importlib-metadata 4.11.3.
> cg, version 22.68.0
> housekeeper, version 3.2.0
> demux, version 5.10.3
> cglims, version 1.3.4

**Pull request merged**